Puntos generales

1. Homogeneizar los nombres de las columnas Todo lower?

In [2]:
import pandas as pd

In [3]:
df_ejemplo=pd.read_csv("HR RAW DATA.csv",index_col=0)
df_ejemplo

Age Attrition BusinessTravel DailyRate                Department  \
0     51        No            NaN    684,0$                       NaN   
1     52        No            NaN    699,0$                       NaN   
2     42        No  travel_rarely    532,0$   Research & Development    
3     47        No  travel_rarely    359,0$                       NaN   
4     46        No            NaN   1319,0$                       NaN   
...   ..       ...            ...       ...                       ...   
1609  36       Yes  travel_rarely    530,0$                       NaN   
1610  45        No     non-travel    805,0$                       NaN   
1611  39        No  travel_rarely    903,0$                       NaN   
1612  36        No     non-travel   1229,0$                       NaN   
1613  46        No            NaN    566,0$                       NaN   

      DistanceFromHome  Education    EducationField  employeecount  \
0                    6          3               NaN              1   
1                    1          4     Life Sciences              1   
2                    4          2  Technical Degree              1   
3                    2          4           Medical              1   
4                    3          3  Technical Degree              1   
...                ...        ...               ...            ...   
1609                 3          1     Life Sciences              1   
1610                 4          2               NaN              1   
1611               -13          5               NaN              1   
1612                 8          4  Technical Degree              1   
1613                 7          2           Medical              1   

     employeenumber  ...  YearsAtCompany  YearsInCurrentRole  \
0             162,0  ...              20                 NaN   
1             259,0  ...              33                 NaN   
2             319,0  ...              22                 NaN   
3               NaN  ...              20                 NaN   
4               NaN  ...              19                 NaN   
...             ...  ...             ...                 ...   
1609          967,0  ...              16                 NaN   
1610          972,0  ...               9                 NaN   
1611            NaN  ...               8                 NaN   
1612          990,0  ...               7                 NaN   
1613            NaN  ...               8                 NaN   

     YearsSinceLastPromotion  YEARSWITHCURRMANAGER  SameAsMonthlyIncome  \
0                         15                    15              19537,0   
1                         11                     9              19999,0   
2                         11                    15              19232,0   
3                          5                     6              17169,0   
4                          2                     8                  NaN   
...                      ...                   ...                  ...   
1609                       3                     7              10325,0   
1610                       0                     8               4447,0   
1611                       0                     7                  NaN   
1612                       0                     7                  NaN   
1613                       0                     7              10845,0   

     DateBirth       Salary                       RoleDepartament  \
0         1972  1000000000$                                   NaN   
1         1971  1000000000$                                   NaN   
2         1981  1000000000$   ManaGER  -  Research & Development    
3         1976  1000000000$                                   NaN   
4         1977  1000000000$                                   NaN   
...        ...          ...                                   ...   
1609      1987  1000000000$                                   NaN   
1610      1978  1000000000$                                   NaN

Columna "AGE"

1. 0 nulos
2. En primera instancia la columna era "object" pero una vez aplicada la función para convertirlo todo a int podremos usar el método discribe() para que nos muestre los estadísticos y sepamos cuál es la edad mínima entre otros. Ésto nos permitirá eliminar la columna "Over18"


In [3]:
from word2number import w2n
#word2numbers es una biblioteca que permite convertir
# palabras que representan números en sus equivalentes numéricos. 
# w2n es una función dentro de esta biblioteca que permite hacer este cambio:
# w2n.word_to_num()


def convertir_a_numero(valor):
    if isinstance(valor, int): # Verifica si el valor ya es un entero, si lo es no hace nada
        return valor
    elif isinstance(valor, str): # Intenta convertir el valor a número
        try: 
            numero = int(valor)
            return numero
        except ValueError:
            try: # aquí es donde se usa la función w2n en el try pq si la conversion falla entonces se usará
                numero = w2n.word_to_num(valor)
                return numero
            except ValueError:
                print(f"Error: No se pudo convertir '{valor}' a número.")
                return None
    else:
        print(f"Error: Tipo de dato no soportado para el valor '{valor}'.")
        return None



ModuleNotFoundError: No module named 'word2number'

In [4]:
df_ejemplo['Age'] = df_ejemplo['Age'].apply(convertir_a_numero) #llamamos a la función
df_ejemplo


NameError: name 'convertir_a_numero' is not defined

In [5]:
df_ejemplo["Age"].describe()

count     1614
unique      54
top         35
freq        84
Name: Age, dtype: object

Columna "Attrition" 
1. No hay nulos y solo hay yes o no.
2. Más adelante veremos cómo trabajarla.

Columna "Business travel"

1. preguntar si los que son NaN es pq no viajan 
2. Qué  % de NaN es relevante para poder eliminar esta columna? Tenemos varios NaN

In [6]:
df_ejemplo["BusinessTravel"].unique()

array([nan, 'travel_rarely', 'travel_frequently', 'non-travel'],
      dtype=object)

Columna "DailyRate"

In [7]:
df_ejemplo["DailyRate"] #aquí hemos mirado la info para la columna

0        684,0$
1        699,0$
2        532,0$
3        359,0$
4       1319,0$
         ...   
1609     530,0$
1610     805,0$
1611     903,0$
1612    1229,0$
1613     566,0$
Name: DailyRate, Length: 1614, dtype: object

In [8]:
df_ejemplo["DailyRate"].isnull().sum()#comprobado si había nulos

0

In [9]:
df_ejemplo.iloc[29:30,:]# hemos identificado un nulo para hacer el test

Age Attrition BusinessTravel DailyRate Department  DistanceFromHome  \
29  34        No     non-travel      nan$     Sales                 10   

    Education EducationField  employeecount employeenumber  ...  \
29          3  Life Sciences              1         1774,0  ...   

    YearsAtCompany  YearsInCurrentRole YearsSinceLastPromotion  \
29              15                 NaN                       0   

    YEARSWITHCURRMANAGER  SameAsMonthlyIncome DateBirth       Salary  \
29                     7               4001,0      1989  1000000000$   

                RoleDepartament NUMBERCHILDREN  RemoteWork  
29   saLEs exEcutIVe  -  Sales             NaN           1  

[1 rows x 41 columns]

In [10]:
df_ejemplo['DailyRate'] =(df_ejemplo['DailyRate'].str.replace('$', '').str.replace(',', '.')) # hemos sacado la unidad de moneda y la coma por punto


In [11]:
import numpy as np

In [12]:
for i in df_ejemplo["DailyRate"]: # hacemos un loop y try except para que nos convierta el object a float, teniendo en cuenta los "nan"(Str) y convertirlos en NaN para que no haya problemas en el momento de medias,etc.
    try:
        df_ejemplo["DailyRate"]=pd.to_numeric(df_ejemplo['DailyRate']).astype(float)
    except:
        df_ejemplo["DailyRate"].replace([i],[np.nan],inplace=True)
   

In [13]:
df_ejemplo.iloc[29:30,:] #testeamos

Age Attrition BusinessTravel  DailyRate Department  DistanceFromHome  \
29  34        No     non-travel        NaN     Sales                 10   

    Education EducationField  employeecount employeenumber  ...  \
29          3  Life Sciences              1         1774,0  ...   

    YearsAtCompany  YearsInCurrentRole YearsSinceLastPromotion  \
29              15                 NaN                       0   

    YEARSWITHCURRMANAGER  SameAsMonthlyIncome DateBirth       Salary  \
29                     7               4001,0      1989  1000000000$   

                RoleDepartament NUMBERCHILDREN  RemoteWork  
29   saLEs exEcutIVe  -  Sales             NaN           1  

[1 rows x 41 columns]

In [14]:
df_ejemplo["DailyRate"].describe() # para ver los valores estadísticos

count    1427.000000
mean      805.413455
std       402.080266
min       103.000000
25%       469.500000
50%       805.000000
75%      1157.000000
max      1499.000000
Name: DailyRate, dtype: float64

Columna "Department"

In [15]:
round(df_ejemplo["Department"].isnull().sum()/df_ejemplo.shape[0],2)#porcentaje de nulos en la columna. Muy elevado a preguntar si vale la pena mantenerla.

0.81

Columna "DistanceFromHome"

In [16]:
df_ejemplo['DistanceFromHome'] = df_ejemplo['DistanceFromHome'].abs() # pasamos los integers de negativos a positivos usando la función abs() 
                                                                        #de panda que calcula el valor absoluto de los integers

df_ejemplo['DistanceFromHome']

0        6
1        1
2        4
3        2
4        3
        ..
1609     3
1610     4
1611    13
1612     8
1613     7
Name: DistanceFromHome, Length: 1614, dtype: int64

_______

COLUMNA 7: Education. Nivel de educación del empleado (generalmente en una escala del 1 al 5).

- ¿A qué corresponde cada número del Nivel Educativo?
    - A lo mejor se debía cambiar por: Estduios básicos - medios , superiores, etc. - > Más comprensible
- Está en float. Debería estar a int? (lo paso debajo)
- No tiene nulos

In [17]:
df_ejemplo["Education"].describe()

count    1614.000000
mean        2.925031
std         1.022357
min         1.000000
25%         2.000000
50%         3.000000
75%         4.000000
max         5.000000
Name: Education, dtype: float64

In [18]:
df_ejemplo["Education"].isnull().sum()

0

In [19]:
df_ejemplo["Education"].value_counts()
# puede ser interesante para ver que hay más gente de nivel de estudios 3 (suponemos medio?) que de nivel 5, qu es el que hay menos (igual es máster o doctorado?)

Education
3    621
4    445
2    314
1    180
5     54
Name: count, dtype: int64

In [20]:
# media_edad_por_educacion = df.groupby('Education')['Age'].mean()
# aplicar cuando tenga la columna edad todo int 

In [21]:
# Para pasar de float a int:
df_ejemplo['Education'] = df_ejemplo['Education'].astype(int)
df_ejemplo['Education']

0       3
1       4
2       2
3       4
4       3
       ..
1609    1
1610    2
1611    5
1612    4
1613    2
Name: Education, Length: 1614, dtype: int64

In [22]:
df_ejemplo['Education'].describe()

count    1614.000000
mean        2.925031
std         1.022357
min         1.000000
25%         2.000000
50%         3.000000
75%         4.000000
max         5.000000
Name: Education, dtype: float64

___

COLUMNA 8. EducationField: El campo de educación del empleado.

- Hay 745 nulos de un total de 1614 registros -> mirar como gestionar éstos nulos
    - Añadirlos a la categoria "other"? (prueba abajo en #)

-  Podemos ver que un alto porcentaje de la platilla viene de la rama de ciencias 

In [23]:
df_ejemplo["EducationField"].describe()

count               869
unique                6
top       Life Sciences
freq                349
Name: EducationField, dtype: object

In [24]:
df_ejemplo["EducationField"].isnull().sum()

745

In [25]:
df_ejemplo["EducationField"]

0                    NaN
1          Life Sciences
2       Technical Degree
3                Medical
4       Technical Degree
              ...       
1609       Life Sciences
1610                 NaN
1611                 NaN
1612    Technical Degree
1613             Medical
Name: EducationField, Length: 1614, dtype: object

In [26]:
df_ejemplo["EducationField"].value_counts()

EducationField
Life Sciences       349
Medical             276
Marketing           104
Technical Degree     69
Other                59
Human Resources      12
Name: count, dtype: int64

In [27]:
# Calculamos el porcentage de cada nivel educativo repeto el total:

conteo_educacion = df_ejemplo["EducationField"].value_counts()

# Calcular el porcentaje
porcentaje_educacion = conteo_educacion / len(df_ejemplo) * 100

# Imprimir el resultado
print("Porcentaje de cada valor en 'EducationField':")
print(porcentaje_educacion)

# Podemos ver que un alto porcentaje de la platilla viene de la rama de ciencias 

Porcentaje de cada valor en 'EducationField':
EducationField
Life Sciences       21.623296
Medical             17.100372
Marketing            6.443618
Technical Degree     4.275093
Other                3.655514
Human Resources      0.743494
Name: count, dtype: float64


In [28]:
# Si cambiamos los valores nulos por other:

#df_ejemplo["EducationField"] = df_ejemplo["EducationField"].fillna("Other")
#df_ejemplo["EducationField"].value_counts()

____

COLUMNA 9. EmployeeCount: Un contador que generalmente es 1 y se usa para contar empleados.
- Eliminar columna por irrelevante

In [29]:
df_ejemplo.drop("employeecount", axis = 1, inplace= True)

In [30]:
df_ejemplo.head(1)

Age Attrition BusinessTravel  DailyRate Department  DistanceFromHome  \
0  51        No            NaN        NaN        NaN                 6   

   Education EducationField employeenumber  EnvironmentSatisfaction  ...  \
0          3            NaN          162,0                        1  ...   

   YearsAtCompany YearsInCurrentRole  YearsSinceLastPromotion  \
0              20                NaN                       15   

   YEARSWITHCURRMANAGER SameAsMonthlyIncome  DateBirth       Salary  \
0                    15             19537,0       1972  1000000000$   

  RoleDepartament  NUMBERCHILDREN  RemoteWork  
0             NaN             NaN         Yes  

[1 rows x 40 columns]

_____

COLUMNA 10. EmployeeNumber: Un número de identificación único para el empleado.
- cambiar a int? -> he intentado cambiat de float a int, pero entonces NaN se convertía en nan (str)
- Hay 534 numeros duplicados
    - Si es un id único, se debería borrar esta columna y hacer otra de nuevo con claves únicas por empleado.
- NaN -> 431
si pasamos a Mysql se puede hacer que se conviertan a Null

- ver que numero están duplicados, por si hay unn empleado que sale 2 veces o si hay dos empleados distintos con el mismo employee number.

In [31]:
df_ejemplo.sample(4)

Age Attrition BusinessTravel  DailyRate Department  DistanceFromHome  \
1446  29        No            NaN      726.0        NaN                29   
1517  47        No  travel_rarely     1482.0        NaN                 5   
352   38        No            NaN      240.0        NaN                 2   
929   24       Yes  travel_rarely      693.0        NaN                 3   

      Education EducationField employeenumber  EnvironmentSatisfaction  ...  \
1446          1  Life Sciences            NaN                        4  ...   
1517          5            NaN          447,0                        4  ...   
352           4  Life Sciences          803,0                        1  ...   
929           2  Life Sciences            NaN                        1  ...   

      YearsAtCompany YearsInCurrentRole  YearsSinceLastPromotion  \
1446               7                NaN                        1   
1517               3                NaN                        1   
352               15                NaN                        4   
929                2                NaN                        2   

      YEARSWITHCURRMANAGER SameAsMonthlyIncome  DateBirth       Salary  \
1446                     6                 NaN       1994  1000000000$   
1517                     1                 NaN       1976  1000000000$   
352                     12              5980,0       1985  1000000000$   
929                      0              4577,0       1999  1000000000$   

     RoleDepartament  NUMBERCHILDREN  RemoteWork  
1446             NaN             NaN           0  
1517             NaN             NaN           1  
352              NaN             NaN        True  
929              NaN             NaN         Yes  

[4 rows x 40 columns]

In [32]:
# Pasamos de str a float -> y se reemplaza la ',' por '.'
df_ejemplo['employeenumber'] = pd.to_numeric(df_ejemplo['employeenumber'].str.replace(',', '.'), errors='coerce')
df_ejemplo['employeenumber'] 

0       162.0
1       259.0
2       319.0
3         NaN
4         NaN
        ...  
1609    967.0
1610    972.0
1611      NaN
1612    990.0
1613      NaN
Name: employeenumber, Length: 1614, dtype: float64

In [33]:
df_ejemplo['employeenumber'].isnull().sum()

431

In [34]:
# cuántos registros hay duplicados?
numeros_duplicados = df_ejemplo[df_ejemplo['employeenumber'].duplicated(keep=False)]
numeros_duplicados['employeenumber'].size

639

In [35]:
#identificamos los números duplicados
duplicados = df_ejemplo[df_ejemplo.duplicated(subset=['employeenumber'], keep=False)]['employeenumber'].unique()
print(duplicados)

#revisar los puntos finales

[  nan  307.  374.  524.  569. 1541.  376.  381.  388.  389.  401.  416.
  430.  438.  446.  448.  454.  458.  476.  483.  488.  500.  534.  586.
  595.  608.  616.  621.  653.  663.  682.  710.  717.  721.  722.  724.
  725.  728.  733.  734.  742.  747.  762.  783.  789.  793.  803.  809.
  823.  842.  913.  927.  941.  956.  966.  967.  972.  990. 1778. 1797.
 1804. 1816. 1821. 1849. 1869. 1898. 1911. 1927. 1947. 1954.  423.  424.
  433.  440.  447.  455.  460.  465.  470.  475.  478.  482.  495.  501.
  502.  507.  517.  522.  523.  525.  526.  530.  544. 1044. 1053. 1069.
 1131. 1135. 1140. 1157. 1160. 1161. 1162. 1185. 1195.]


In [36]:
#comprobamos algunos registros

fila_con_307 = df_ejemplo[df_ejemplo['employeenumber'] == 307]
fila_con_307

#Nota: Vemos que RemoteWork 1 equivale a YES

Age Attrition BusinessTravel  DailyRate Department  DistanceFromHome  \
82    38        No            NaN      119.0        NaN                 3   
1474  38        No            NaN      119.0        NaN                 3   

      Education EducationField  employeenumber  EnvironmentSatisfaction  ...  \
82            3  Life Sciences           307.0                        1  ...   
1474          3  Life Sciences           307.0                        1  ...   

      YearsAtCompany YearsInCurrentRole  YearsSinceLastPromotion  \
82                16                NaN                        5   
1474              16                NaN                        5   

      YEARSWITHCURRMANAGER SameAsMonthlyIncome  DateBirth       Salary  \
82                      13                 NaN       1985  1000000000$   
1474                    13                 NaN       1985  1000000000$   

     RoleDepartament  NUMBERCHILDREN  RemoteWork  
82               NaN             NaN           1  
1474             NaN             NaN         Yes  

[2 rows x 40 columns]

In [37]:
lista_duplicados2= []

for numero in duplicados:
    lista_duplicados2.append(df_ejemplo[df_ejemplo['employeenumber'] == numero])

In [38]:
# lo visualizamos en un nuevo dataFrame
df_resultado = pd.concat(lista_duplicados2)
df_resultado

Age Attrition     BusinessTravel  DailyRate                Department  \
82    38        No                NaN      119.0                       NaN   
1474  38        No                NaN      119.0                       NaN   
84    55        No      travel_rarely      452.0                       NaN   
1476  55        No      travel_rarely      452.0                       NaN   
89    31        No                NaN      480.0                       NaN   
...   ..       ...                ...        ...                       ...   
1494  27        No      travel_rarely      199.0                       NaN   
1008  35        No  travel_frequently        NaN                       NaN   
1495  35        No  travel_frequently        NaN                       NaN   
1010  45        No                NaN     1457.0   Research & Development    
1497  45        No                NaN     1457.0   Research & Development    

      DistanceFromHome  Education EducationField  employeenumber  \
82                   3          3  Life Sciences           307.0   
1474                 3          3  Life Sciences           307.0   
84                   1          3        Medical           374.0   
1476                 1          3        Medical           374.0   
89                  30          2            NaN           524.0   
...                ...        ...            ...             ...   
1494                 6          3            NaN          1162.0   
1008                 4          4          Other          1185.0   
1495                 4          4          Other          1185.0   
1010                 7          3        Medical          1195.0   
1497                 7          3        Medical          1195.0   

      EnvironmentSatisfaction  ...  YearsAtCompany YearsInCurrentRole  \
82                          1  ...              16                NaN   
1474                        1  ...              16                NaN   
84                          4  ...              36                NaN   
1476                        4  ...              36                NaN   
89                          2  ...              13                NaN   
...                       ...  ...             ...                ...   
1494                        4  ...               4                2,0   
1008                        4  ...               2                NaN   
1495                        4  ...               2                NaN   
1010                        1  ...               3                NaN   
1497                        1  ...               3                NaN   

      YearsSinceLastPromotion  YEARSWITHCURRMANAGER SameAsMonthlyIncome  \
82                          5                    13                 NaN   
1474                        5                    13                 NaN   
84                          4                    13             19045,0   
1476                        4                    13             19045,0   
89                          3                    12              4306,0   
...                       ...                   ...                 ...   
1494                        2                     2              2539,0   
1008                        2                     2              2376,0   
1495                        2                     2              2376,0   
1010                        0                     2                 NaN   
1497                        0                     2                 NaN   

      DateBirth       Salary                                  RoleDepartament  \
82         1985  1000000000$                                              NaN   
1474       1985  1000000000$                                              NaN   
84         1968  1000000000$                                              NaN   
1476       1968  1000000000$                                              NaN   
89         1992  1000000000$                                        

Vemos que hay 104 filas que se tienen que eliminar

__________

COLUMNA 11. EnvironmentSatisfaction: Nivel de satisfacción del empleado en relación con su entorno de trabajo. Con valores que estan comprendidos entre el 1 y el 4, siendo el 4 el nivel de máxima satisfacción.

- Hay 100 registros que no van del 1 al 4 -> # Convertimos los valores no validos (superiores a 4) en NaN -> está abajo desactivado con #  abajo.
- No hay nulos


In [39]:
df_ejemplo['EnvironmentSatisfaction']

0        1
1        3
2        3
3        1
4        1
        ..
1609     3
1610     3
1611    13
1612     1
1613     4
Name: EnvironmentSatisfaction, Length: 1614, dtype: int64

In [40]:
# Mostrar los valores mayores a 4
resultados_superiores_a_4 = df_ejemplo[df_ejemplo['EnvironmentSatisfaction'] > 4]['EnvironmentSatisfaction']
resultados_superiores_a_4.count()

100

In [41]:
resultados_superiores_a_4.isnull().sum()

0

In [42]:
# Convertimos los valores no validos (superiores a 4) en NaN: 

# df_ejemplo.loc[df_ejemplo['EnvironmentSatisfaction'] > 4, 'EnvironmentSatisfaction'] = np.nan

# df_ejemplo['EnvironmentSatisfaction']

_____

COLUMNA 12. Gender: El género del empleado. Donde 0 corresponde con "hombre" y 1 con "mujer".

- Hay un 60,1% de hombres respecto a un 39.8% de mujeres
- cambio los valores 0,  1, por male, female para hacerlo más comprensible
- No hay nulos

In [43]:
df_ejemplo['Gender'].value_counts()

Gender
0    971
1    643
Name: count, dtype: int64

In [44]:
df_ejemplo['Gender'].isnull().sum()

0

In [45]:
#Calcular porcentaje:

conteo_genero = df_ejemplo['Gender'].value_counts()

porcentaje_genero = conteo_genero / len(df_ejemplo) * 100
porcentaje_genero

Gender
0    60.16109
1    39.83891
Name: count, dtype: float64

In [46]:
#cambiar los valores 0,  1, por male, female para hacerlo más comprensible

df_ejemplo['Gender'] = df_ejemplo['Gender'].replace({0: 'male', 1: 'female'})
df_ejemplo['Gender']

0         male
1         male
2         male
3       female
4       female
         ...  
1609      male
1610      male
1611      male
1612      male
1613      male
Name: Gender, Length: 1614, dtype: object

______

COLUMNA 13. HourlyRate: La tarifa por hora del empleado.

- NO hay nulos
- Pasamos de srt a int. 
- La media cobra por hora 62.59 

In [47]:
df_ejemplo['HourlyRate'].isnull().sum()

0

In [48]:
# Pasamos el valor a integer. 
# ponemos errors='coerce' ya que hay valores que no se puede transforar, como es el caso de 'Not Available'. Se convertiá a NaN

df_ejemplo['HourlyRate'] = pd.to_numeric(df_ejemplo['HourlyRate'], errors='coerce')

# Convertir la columna a tipo entero, tratando los NaN como 0
df_ejemplo['HourlyRate'] = df_ejemplo['HourlyRate'].fillna(0).astype(int)

In [49]:
df_ejemplo['HourlyRate'].isnull().sum()

0

In [50]:
df_ejemplo['HourlyRate']

0       51
1       65
2       58
3       82
4       45
        ..
1609    51
1610    57
1611    41
1612    84
1613    75
Name: HourlyRate, Length: 1614, dtype: int64

In [51]:
df_ejemplo['HourlyRate'].describe()

count    1614.000000
mean       62.591698
std        24.525022
min         0.000000
25%        46.000000
50%        64.000000
75%        83.000000
max       100.000000
Name: HourlyRate, dtype: float64

____

COLUMNA 14. JobInvolvement: Nivel de implicación del empleado en su trabajo.
- Que significa? Quien lo valora? El mánager/jefe? Es una auto-evaluación?
- Se piede relacionar con otras columnas? Ej: salario / número de horas que se trabajan, etc...

In [52]:
df_ejemplo['JobInvolvement'].value_counts()

JobInvolvement
3    955
2    406
4    164
1     89
Name: count, dtype: int64

In [53]:
df_ejemplo['JobInvolvement'].size

1614

In [54]:
df_ejemplo['JobInvolvement'].isnull().sum()

0

______

COLUMNA 15: JobLevel - Nivel Jerárquico del empleado en la empresa

-tipo de dato INT

-No hay valores nulos

In [4]:
df_ejemplo['JobLevel'].unique()

array([5, 4, 3, 2, 1])

In [6]:
df_ejemplo['JobLevel'].isnull().sum()

0

COLUMNA 16: JobRole -El rol o puesto de trabajo del empleado.

-No hay nulos

-Tipo de dato 'object'

-Cambie la infromacion a letras minusculas (estaba en formato messenger)

In [ ]:
df_ejemplo['JobRole']= [elemento.lower() for elemento in df_ejemplo['JobRole']]

In [ ]:
df_ejemplo['JobRole'].value_counts()

COLUMNA 17: JobSatisfaction - Nivel de satisfacción del empleado con su trabajo.

-Tipo de dato: int64

-No hay valores nulos

-Dejo registrada la media de satisfaccion laboral

In [7]:
df_ejemplo['JobSatisfaction'].unique()

array([3, 4, 1, 2])

In [8]:
df_ejemplo['JobSatisfaction'].value_counts()

4    514
3    481
1    317
2    302
Name: JobSatisfaction, dtype: int64

In [9]:
df_ejemplo['JobSatisfaction'].describe(include='object')

count    1614.000000
mean        2.738538
std         1.106163
min         1.000000
25%         2.000000
50%         3.000000
75%         4.000000
max         4.000000
Name: JobSatisfaction, dtype: float64

COLUMNA 18: MaritalStatus - El estado civil del empleado (por ejemplo, "Single", "Married", etc.).

-Tipo de dato: 

-Datos nulos: 651. NaN pueden ser floats o str 

-Tenemos Divorce escrito de dos formas diferentes. Divorced/divorced. Lo unifico en Divorced
Ver si lo modificamos todo a minusculas
-Married idem anterior

In [10]:
df_ejemplo['MaritalStatus'].isnull().sum()

651

In [11]:
df_ejemplo['MaritalStatus']= df_ejemplo['MaritalStatus'].replace('divorced', 'Divorced')

In [12]:
df_ejemplo['MaritalStatus']= df_ejemplo['MaritalStatus'].replace('Marreid', 'Married')

In [13]:
df_ejemplo['MaritalStatus'].value_counts()

Married     439
Single      325
Divorced    199
Name: MaritalStatus, dtype: int64

COLUMNA 19: MonthlyIncome - Ingresos mensuales del empleado.

-Datos nulos: 843

-Tipo de dato: object

In [18]:
df_ejemplo['MonthlyIncome'].isnull().sum()

843

In [ ]:
df_ejemplo["MonthlyIncome"]= df_ejemplo["MonthlyIncome"].str.replace(',','.')

In [20]:
df_ejemplo["MonthlyIncome"]=pd.to_numeric(df_ejemplo["MonthlyIncome"]).astype(float)

COLUMNA 20: MonthlyRate - Tasa mensual del empleado.

 _ significa la tasa mensual de salario o sueldo que recibe un empleado por su trabajo en una empresa. Es el monto específico que se paga a los empleados en forma regular cada mes. Esta información se suele registrar en una tabla de empleados para llevar un registro de los salarios y compensaciones que se les otorgan. La tasa mensual puede variar dependiendo del puesto, la experiencia, las habilidades y otros factores relevantes.

-tipo de dato: int 

-No hay nulos

Hay mucha diferencia entre el minimo y el máximo

In [22]:
df_ejemplo['MonthlyRate'].describe()

count     1614.000000
mean     14284.495663
std       7110.414585
min       2094.000000
25%       8001.000000
50%      14248.500000
75%      20364.000000
max      26999.000000
Name: MonthlyRate, dtype: float64

COLUMNA 21: NUMCOMPANIESWORKED - Número de compañías en las que el empleado ha trabajado.

-tipo de dato: Int

-No hay nulos 

-Cambiamos el tamaño del nombre de la columna, mismo patron que el resto de las columnas

In [25]:
df_ejemplo['NUMCOMPANIESWORKED'].unique()

array([7, 0, 1, 3, 2, 4, 8, 9, 5, 6])

In [26]:
df_ejemplo['NUMCOMPANIESWORKED'].value_counts()

1    573
0    226
3    169
4    157
2    156
7     84
6     73
5     66
9     59
8     51
Name: NUMCOMPANIESWORKED, dtype: int64

Over18 - Indica si el empleado es mayor de 18 años.

-valores nulos: 901

-Esta columna no tiene sentido


In [27]:
df_ejemplo['Over18'].isnull().sum()

901

In [28]:
df_ejemplo['Over18'].unique()

array(['Y', nan], dtype=object)

______

COLUMNA 23: OverTime

solo 3 tipos de valores, 40% Nan

In [ ]:
df_ejemplo['OverTime'].unique()

In [ ]:
df_ejemplo['OverTime'].value_counts()

In [ ]:
df_ejemplo['OverTime'].isnull().sum()

COLUMNA 24: PercentSalaryHike	

muchos porcentajes, ningún null

In [ ]:
df_ejemplo['PercentSalaryHike'].unique()

In [ ]:
df_ejemplo['PercentSalaryHike'].value_counts()

In [ ]:
df_ejemplo['PercentSalaryHike'].describe()

In [ ]:
df_ejemplo['PercentSalaryHike'].isnull().sum()

COLUMNA 25: PerformanceRating

Pocos null, hay que cambiar los float por int, solo valores 3 y 4

In [ ]:
df_ejemplo['PerformanceRating'].unique()

In [ ]:
df_ejemplo['PerformanceRating'].value_counts()

In [ ]:
df_ejemplo['PerformanceRating'].describe()

In [ ]:
df_ejemplo['PerformanceRating']=df_ejemplo['PerformanceRating'].str.replace(',','.')
df_ejemplo['PerformanceRating']=df_ejemplo['PerformanceRating'].astype(float)

In [ ]:
df_ejemplo['PerformanceRating'].describe()

In [ ]:
df_ejemplo['PerformanceRating'].isnull().sum()

COLUMNA 26: RelationshipSatisfaction

este aprece que está bien

In [ ]:
df_ejemplo['RelationshipSatisfaction'].unique()

In [ ]:
df_ejemplo['RelationshipSatisfaction'].value_counts()

In [ ]:
df_ejemplo['RelationshipSatisfaction'].describe()

In [ ]:
df_ejemplo['RelationshipSatisfaction'].isnull().sum()

COLUMNA 27: StandardHours	

solo 80 y nan. Podemos eliminarla

In [ ]:
df_ejemplo['StandardHours'].unique()

In [ ]:
df_ejemplo['StandardHours'].isnull().sum()

COLUMNA 28: StockOptionLevel

podríamos cambiar el nombre de la columna para que sea mas entendible

In [ ]:
df_ejemplo['StockOptionLevel'].unique()

In [ ]:
df_ejemplo['StockOptionLevel'].value_counts()

In [ ]:
df_ejemplo['StockOptionLevel'].describe()

In [ ]:
df_ejemplo['StockOptionLevel'].isnull().sum()

COLUMNA 29: TOTALWORKINGYEARS

hay que cambiar el nombre de la col por minus o title, cambiar los float por int

In [ ]:
df_ejemplo['TOTALWORKINGYEARS'].unique()

In [ ]:
df_ejemplo['TOTALWORKINGYEARS'].describe()

In [ ]:
df_ejemplo['TOTALWORKINGYEARS']=df_ejemplo['TOTALWORKINGYEARS'].str.replace(',','.')
df_ejemplo['TOTALWORKINGYEARS']=df_ejemplo['TOTALWORKINGYEARS'].astype(float)

In [ ]:
df_ejemplo['TOTALWORKINGYEARS'].describe()

In [ ]:
df_ejemplo['TOTALWORKINGYEARS'].isnull().sum()

In [ ]:
df_ejemplo.rename(columns={'TOTALWORKINGYEARS':'TotalWorkingYears'})

COLUMNA 30: TrainingTimesLastYear

Todo parece coherente

In [ ]:
df_ejemplo['TrainingTimesLastYear'].unique()

In [ ]:
df_ejemplo['TrainingTimesLastYear'].value_counts()

In [ ]:
df_ejemplo['TrainingTimesLastYear'].isnull().sum()

______

Columna "YearsSinceLastPromotion"

1. no hay nulos y una columna de integers

In [55]:
df_ejemplo["YearsSinceLastPromotion"].isnull().sum()

df_ejemplo["YearsSinceLastPromotion"].info()

<class 'pandas.core.series.Series'>
Index: 1614 entries, 0 to 1613
Series name: YearsSinceLastPromotion
Non-Null Count  Dtype
--------------  -----
1614 non-null   int64
dtypes: int64(1)
memory usage: 25.2 KB


Columna "YEARSWITHCURRMANAGER"

In [56]:
df_ejemplo["YEARSWITHCURRMANAGER"].info()

<class 'pandas.core.series.Series'>
Index: 1614 entries, 0 to 1613
Series name: YEARSWITHCURRMANAGER
Non-Null Count  Dtype
--------------  -----
1614 non-null   int64
dtypes: int64(1)
memory usage: 25.2 KB


Columna "SameAsMonthlyIncome"

1. En principio es object y convertirla a float
2. Además es una columna con nulos, representan un 52% (nulos total 843)
3. compararla con "MonthlyIncome"
4. a eliminar pq es lo mismo que en MonthlyIncome

In [57]:
df_ejemplo["SameAsMonthlyIncome"].info()

<class 'pandas.core.series.Series'>
Index: 1614 entries, 0 to 1613
Series name: SameAsMonthlyIncome
Non-Null Count  Dtype 
--------------  ----- 
771 non-null    object
dtypes: object(1)
memory usage: 25.2+ KB


In [58]:
df_ejemplo['SameAsMonthlyIncome'] = df_ejemplo['SameAsMonthlyIncome'].replace(',', '.') #para poder convertir de obj a float primero hay 
                                                                                        #q sustituir las comas por puntos. Los float van con puntos
df_ejemplo['SameAsMonthlyIncome'] = df_ejemplo['SameAsMonthlyIncome'].astype(float) #convertimos a float
df_ejemplo['SameAsMonthlyIncome']

ValueError: could not convert string to float: '19537,0'

Columna "DateBirth"

1. Cambiar nombre de la columna a "Yearbirth"



In [ ]:
df_ejemplo["DateBirth"].info()

<class 'pandas.core.series.Series'>
Index: 1614 entries, 0 to 1613
Series name: DateBirth
Non-Null Count  Dtype
--------------  -----
1614 non-null   int64
dtypes: int64(1)
memory usage: 89.8 KB


Columna "Salary"

1. object convertirlo a float
2. eliminar el signo de dólares (referenciarlo en el nombre de la columna)
3. solo hay un elemento en la columna - no tiene mucho sentido candidata a ser eliminada y rehacerla con calculos 


In [ ]:
df_ejemplo["Salary"].info()

<class 'pandas.core.series.Series'>
Index: 1614 entries, 0 to 1613
Series name: Salary
Non-Null Count  Dtype 
--------------  ----- 
1614 non-null   object
dtypes: object(1)
memory usage: 89.8+ KB


In [ ]:
df_ejemplo["Salary"].nunique()

1

In [ ]:
df_ejemplo['Salary'] = df_ejemplo['Salary'].str.replace('$', '') #sacamos el símbolo de la currency de la columna para poder convertir a float

# Convertir los valores a tipo float 
df_ejemplo['Salary'] = df_ejemplo['Salary'].astype(float)

Columna "RoleDepartament"

1. % elevado de null un 81%, candidata a ser eliminada


In [ ]:
df_ejemplo["RoleDepartament"].isnull().sum()

1312

In [ ]:
df_ejemplo["RoleDepartament"].unique()

array([nan, ' ManaGER  -  Research & Development ',
       ' MANAger  -  Research & Development ',
       ' heaLtHcArE repResENtATiVe  -  Research & Development ',
       ' saLEs exEcutIVe  -  Sales ', ' Sales ExECuTIVe  -  Sales ',
       ' heALthCArE RePResEntaTIVe  -  Research & Development ',
       ' lABORAtOry tECHnIcIan  -  Research & Development ',
       ' ManufactUrINg DiReCTOr  -  Research & Development ',
       ' SaleS eXEcUtIVE  -  Sales ',
       ' HEaLthcaRe rEprEsEnTAtiVe  -  Research & Development ',
       ' rESeArCH sCientIST  -  Research & Development ',
       ' healTHCAre rePREseNtATiVE  -  Research & Development ',
       ' mANaGer  -  Research & Development ',
       ' HEAltHCaRe REPreseNTatIve  -  Research & Development ',
       ' SaLeS eXECUTivE  -  Sales ',
       ' reSeARCH SCIEnTiST  -  Research & Development ',
       ' ReSearch DiRecTOR  -  Research & Development ',
       ' resEaRCh SciEnTist  -  Research & Development ',
       ' hUmaN REsoUrCes  -  H

In [ ]:
df_ejemplo["RoleDepartament"].value_counts()

RoleDepartament
 MaNAgeR  -  Sales                                        2
 ManaGER  -  Research & Development                       1
 ReseaRch scIENTisT  -  Research & Development            1
 ManufacTURInG DIRECtOR  -  Research & Development        1
 hEalthCaRe reprEseNTaTiVe  -  Research & Development     1
                                                         ..
 saLES eXEcUTiVE  -  Sales                                1
 mANUfacTURiNG dIRectOR  -  Research & Development        1
 huMAn ResOurces  -  Human Resources                      1
 HUMAN ResoURCeS  -  Human Resources                      1
 sAleS EXECUtIvE  -  Sales                                1
Name: count, Length: 301, dtype: int64

Columna "NUMBERCHILDREN"

1. no hay valores, mejor eliminarla

In [ ]:
df_ejemplo["NUMBERCHILDREN"].info()

<class 'pandas.core.series.Series'>
Index: 1614 entries, 0 to 1613
Series name: NUMBERCHILDREN
Non-Null Count  Dtype  
--------------  -----  
0 non-null      float64
dtypes: float64(1)
memory usage: 89.8 KB


Columna "RemoteWork"

1. hay 5 valores unicos: 0,1,yes,false,true; no sabemos qué es 0 y 1


In [ ]:
df_ejemplo["RemoteWork"].value_counts()


RemoteWork
1        360
True     345
0        309
False    305
Yes      295
Name: count, dtype: int64

In [ ]:
df_ejemplo["RemoteWork"].info()

<class 'pandas.core.series.Series'>
Index: 1614 entries, 0 to 1613
Series name: RemoteWork
Non-Null Count  Dtype 
--------------  ----- 
1614 non-null   object
dtypes: object(1)
memory usage: 25.2+ KB


COLUMNA WORKLIFEBALANCE

In [ ]:
df_ejemplo["WORKLIFEBALANCE"].info()

<class 'pandas.core.series.Series'>
Index: 1614 entries, 0 to 1613
Series name: WORKLIFEBALANCE
Non-Null Count  Dtype 
--------------  ----- 
1506 non-null   object
dtypes: object(1)
memory usage: 25.2+ KB


In [ ]:
df_ejemplo["WORKLIFEBALANCE"].value_counts()

WORKLIFEBALANCE
3,0    913
2,0    359
4,0    155
1,0     79
Name: count, dtype: int64

In [ ]:
df_ejemplo["WORKLIFEBALANCE"].info()

In [ ]:
df_ejemplo["WORKLIFEBALANCE"].isnull().sum()

108

In [ ]:
df_ejemplo['WORKLIFEBALANCE'] = df_ejemplo['WORKLIFEBALANCE'].str.replace(',', '.') #sacamos el símbolo de la currency de la columna para poder convertir a float

# Convertir los valores a tipo float 
df_ejemplo['WORKLIFEBALANCE'] = df_ejemplo['WORKLIFEBALANCE'].astype(float)

Columna YearsAtCompany

In [ ]:
df_ejemplo["YearsAtCompany"].info()

<class 'pandas.core.series.Series'>
Index: 1614 entries, 0 to 1613
Series name: YearsAtCompany
Non-Null Count  Dtype
--------------  -----
1614 non-null   int64
dtypes: int64(1)
memory usage: 25.2 KB


In [ ]:
df_ejemplo["YearsAtCompany"].value_counts()

YearsAtCompany
5     208
1     171
3     141
2     141
10    133
7     115
4     114
8     106
9      94
6      78
0      44
11     36
20     29
13     26
15     21
14     19
22     17
12     15
18     15
16     14
21     14
19     12
17      9
24      7
25      5
33      5
26      4
32      3
27      3
31      3
36      3
29      2
23      2
40      2
34      1
37      1
30      1
Name: count, dtype: int64

Columna YearsInCurrentRole
1. muchos nulos quizás conveniente eliminarla

In [ ]:
df_ejemplo["YearsInCurrentRole"].info()


<class 'pandas.core.series.Series'>
Index: 1614 entries, 0 to 1613
Series name: YearsInCurrentRole
Non-Null Count  Dtype 
--------------  ----- 
34 non-null     object
dtypes: object(1)
memory usage: 25.2+ KB


In [ ]:
df_ejemplo["YearsInCurrentRole"].isnull().sum()

1580